File: 03.CPP/0643.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    double findMaxAverage(std::vector<int>& nums, int k) {
        int n = nums.size();

        int windowSum = 0;
        for (int i = 0; i < k; i++) {
            windowSum += nums[i];
        }

        int maxSum = windowSum;
        for (int i = k; i < n; i++) {
            windowSum += nums[i] - nums[i - k];

            if (windowSum > maxSum) {
                maxSum = windowSum;
            }
        }

        double maxAvg = (double) maxSum / k;
        return maxAvg;
    }
};

int main() {
    Solution sol;
    
    std::vector<int> nums = {1, 12, -5, -6, 50, 3};
    int k = 4;
    double result = sol.findMaxAverage(nums, k);

    std::cout << "Result: " << result << "\n";
    return 0;
}

File: 01.Python/0643.py
from typing import List

class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        n = len(nums)

        if k > n:
            return -1.0

        windowSum = 0
        for i in range(k):
            windowSum += nums[i]
        
        maxSum = windowSum
        for i in range(k, n):
            windowSum += nums[i] - nums[i - k]

            if windowSum > maxSum:
                maxSum = windowSum

        return maxSum / k
    
sol = Solution()

nums = [1, 12, -5, -6, 50, 3]
k = 4

result = sol.findMaxAverage(nums, k)
print(f"The maximum average of {k}-length subarray in {nums} is {result}")